File: github/src/main/java/com/github/GitHubClient.java
package com.github;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLConnection;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by swap_i on 21/12/14.
 */
public class GitHubClient {

    private static final Logger L = LoggerFactory.getLogger(GitHubClient.class);
    private static final Pattern CELL_PATTERN = Pattern.compile("<rect class=\"day\" .+? data-count=\"(\\d+?)\" data-date=\"(\\d{4}-\\d{2}-\\d{2})\"/>", Pattern.DOTALL);
    private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd");

    public List<DayActivityFromServer> getUserActivity(String username) throws PageParseException {

        try {
            URLConnection connection = new URL("https://github.com/" + username).openConnection();

            connection.connect();

            StringBuilder builder = new StringBuilder();
            InputStreamReader reader = new InputStreamReader(connection.getInputStream());
            try {
                int readed;
                char[] buffer = new char[1024];
                while ((readed = reader.read(buffer)) != -1) {
                    builder.append(buffer, 0, readed);
                }
            } finally {
                reader.close();
            }

            String resultStream = builder.toString();
            Matcher matcher = CELL_PATTERN.matcher(resultStream);

            List<DayActivityFromServer> list = new ArrayList<>();
            while (matcher.find()) {

                int count = Integer.parseInt(matcher.group(1));
                Date date = DATE_FORMAT.parse(matcher.group(2));

                DayActivityFromServer activity = new DayActivityFromServer();
                activity.setDate(date);
                activity.setActivityCount(count);
                list.add(activity);
            }

			if (list.isEmpty()) {
				throw new PageParseException("Can't find any activity on page");
			}

            return list;

        } catch (IOException e) {
            L.error("Error", e);

        } catch (ParseException e) {
            L.error("Error", e);
        }

        return new ArrayList<>();
    }

	public class PageParseException extends Exception {

		public PageParseException(String s) {
			super(s);
		}

	}

}

File: github/src/main/java/com/github/GitHubClient.kt
package com.github

import hu.akarnokd.rxjava3.bridge.RxJavaBridge
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import kotlinx.coroutines.rx2.rxSingle
import okhttp3.OkHttpClient
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.regex.Pattern

class GitHubClient {

    fun getUserActivity(username: String): Observable<OneDayActivityFromServer> =
        getPageSource(URL("https://github.com/$username"))
            .flatMapObservable { pageSource -> parseDayActivity(pageSource) }
            .switchIfEmpty(Observable.error(PageParseException("Can't find any activity on page")))

    private fun getPageSource(url: URL): Single<String> =
        RxJavaBridge.toV3Single(
            rxSingle {
                GetWebPageSource(OkHttpClient()).invoke(url)
            }
        )

    private fun parseDayActivity(pageSource: String): Observable<OneDayActivityFromServer> =
        Observable.create { emitter ->
            val matcher = CELL_PATTERN.matcher(pageSource)
            while (matcher.find()) {
                val count = Integer.parseInt(matcher.group(1))
                val date = DATE_FORMAT.parse(matcher.group(2))
                emitter.onNext(OneDayActivityFromServer(date, count))
            }
            emitter.onComplete()
        }

    inner class PageParseException(s: String) : RuntimeException(s)

    companion object {
        private val CELL_PATTERN = Pattern.compile(
            "<rect class=\"day\" .+? data-count=\"(\\d+?)\" data-date=\"(\\d{4}-\\d{2}-\\d{2})\"/>",
            Pattern.DOTALL
        )
        private val DATE_FORMAT = SimpleDateFormat("yyyy-MM-dd", Locale.ROOT)
    }

}
